xarray/core/rolling.py

```diff
@@ -389,7 +389,7 @@ def _bottleneck_reduce(self, func, **kwargs):
                 valid = (slice(None),) * axis + (slice(-shift, None),)
             padded = padded.pad({self.dim[0]: (0, -shift)}, mode="constant")
 
-        if isinstance(padded.data, dask_array_type):
+        if is_duck_dask_array(padded.data):
             raise AssertionError("should not be reachable")
```

Comment:
@pydata/xarray: why is that here? If it is never reachable, it should be fine to just not do the branching (and if it is necessary, it should be fine to use `assert not is_duck_dask_array(padded.data), "should not be reachable"`).

Also, if we raise the code below will never be reached, so we should be able to remove it.

Reply:
I think it's from https://github.com/pydata/xarray/pull/3040 . We would like to use this code path but there are bugs IIRC

Reply:
then I guess we should raise a `NotImplementedError` or a `ValueError` instead of a `AssertionError` and improve the error message?

Reply:
We could comment it out and add a TODO note. The `if` condition in Line 420 makes sure that this is never reached.